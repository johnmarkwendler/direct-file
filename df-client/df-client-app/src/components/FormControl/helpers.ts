import { ConcretePath } from '@irs/js-factgraph-scala';
import { Path } from '../../fact-dictionary/Path.js';

export function buildFormControlId<Subfield extends string>(concretePath: ConcretePath, subfield?: Subfield) {
  // This feels silly, but provides some security if we ever decide on a different naming scheme, for some reason
  const optionalSubfield = subfield ? `__` + subfield : ``;
  return `id-${concretePath}${optionalSubfield}` as const;
}

export function buildControlErrorId<Subfield extends string>(
  controlId: ReturnType<typeof buildFormControlId>,
  subfield?: Subfield
) {
  const optionalSubfield = subfield ? `__` + subfield : ``;
  return `${controlId}${optionalSubfield}__error-msg` as const;
}

export function buildHintId<Subfield extends string>(
  controlId: ReturnType<typeof buildFormControlId> | string,
  subfield?: Subfield
) {
  const optionalSubfield = subfield ? `__` + subfield : ``;
  return `${controlId}${optionalSubfield}__hint` as const;
}

export function buildHintKey<Subfield extends string>(path?: Path, subField?: Subfield) {
  if (path) {
    const optionalSubfield = subField ? `.${subField}` : ``;
    return `info./info${path}${optionalSubfield}.helpText.hint`; //Translation.tsx will add the .text to the final key
  } else {
    return ``;
  }
}

// Edge case: Email field is readonly and uses a DFmodal as the hint.
export function buildReadonlyHintKey(path?: Path) {
  if (path) {
    return `info.${path}.readOnlyField`; // DFModal component will add the .helpText.modals to the final key
  } else {
    return ``;
  }
}

/* Return the original string, and set empty boolean
 */
export function sanitizeString(inputString: string) {
  // Trims trailing and leading whitespaces, collapses adjacent spaces
  const value = inputString.trim().replace(/ +/g, ` `);
  return { value, isEmpty: value === `` };
}
