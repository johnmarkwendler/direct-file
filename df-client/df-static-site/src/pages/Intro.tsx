import { ReactNode, useContext } from 'react';
import { Helmet } from 'react-helmet-async';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';

import { CommonContentDisplay } from '@irs/df-common';
import { Icon } from '@trussworks/react-uswds';

import { ChatIcon, DFAccordion, Heading, Translation as T } from '../components/index.js';
import { PilotPhaseContext } from '../layouts/Providers.js';
import { baseRouterPath } from '../constants.js';

const Section = ({ children, id = undefined }: { children: ReactNode | ReactNode[]; id?: string }) => (
  <section id={id}>
    <div className='grid-container'>{children}</div>
  </section>
);

const Content = ({ i18nKey, noBody = true }: { i18nKey: string; noBody?: boolean }) => (
  <CommonContentDisplay i18nKey={i18nKey} noBody={noBody} TranslationComponent={T} />
);

const BlueCheckedItems = ({ children }: { children: ReactNode[] }) => (
  <ul className='grid-row usa-icon-list'>
    {children.map((child, index) => (
      <li key={index} className='grid-col-6 usa-icon-list__item'>
        <div className='text-blue usa-icon-list__icon' aria-hidden='true'>
          <Icon.Check />
        </div>
        <div className='usa-icon-list__content'>{child}</div>
      </li>
    ))}
  </ul>
);

const Intro = () => {
  const { t } = useTranslation(`translation`);
  const phase = useContext(PilotPhaseContext);
  const NEXT = `/state`;

  return (
    <>
      <Helmet>
        <title>{t(`pages.Landing.pageTitle`)}</title>
      </Helmet>
      <Section id='welcome'>
        <Heading splash>
          {(phase?.showInvitationalHero || phase?.showAfterDeadlineInvitationalHero) && (
            <Content i18nKey='pages.Landing.heading.invitational' />
          )}
          {phase?.showThankYouHero && <Content i18nKey='pages.Landing.heading.thank-you' />}
        </Heading>
        <div className='grid-row'>
          <div className='grid-col-12 desktop:grid-col-6 tablet:grid-col-6'>
            <p className='lead-in'>
              {phase?.showInvitationalHero && <Content i18nKey='pages.Landing.section.welcome.invitational.lead-in' />}
              {phase?.showAfterDeadlineInvitationalHero && (
                <Content i18nKey='pages.Landing.section.welcome.invitational.lead-in-after-deadline' />
              )}
            </p>
            <p className='lead-in margin-bottom-1'>
              <Content i18nKey='pages.Landing.section.welcome.invitational.checklist.header' />
            </p>
            <BlueCheckedItems>
              <Content i18nKey='pages.Landing.section.welcome.invitational.checklist.first' />
              <Content i18nKey='pages.Landing.section.welcome.invitational.checklist.second' />
              <Content i18nKey='pages.Landing.section.welcome.invitational.checklist.third' />
              <Content i18nKey='pages.Landing.section.welcome.invitational.checklist.fourth' />
            </BlueCheckedItems>
            <p className='lead-in'>
              <Content i18nKey='pages.Landing.section.welcome.invitational.second-paragraph' />
            </p>
          </div>

          {phase?.showLandingPageImage && (
            <div className='grid-col-12 desktop:grid-col-6 tablet:grid-col-6 margin-top-5 hidden-on-mobile'>
              <img alt='' className='padding-left-05' src={`${baseRouterPath}/imgs/blueperson_laptop.png`} />
            </div>
          )}
        </div>

        <div className='grid-row grid-gap'>
          <div className='grid-col-12 desktop:grid-col-8 margin-bottom-2'>
            {phase?.enableScreener && (
              <>
                <Link className='usa-button' to={NEXT}>
                  {t(`pages.Landing.button.text`)}
                </Link>
              </>
            )}
          </div>
        </div>
      </Section>

      {phase?.showDataImportContent && (
        <Section id='data-import'>
          <h2>
            <Content i18nKey='pages.Landing.section.dataImport.heading' />
          </h2>
          <div className='grid-row content-container'>
            <p>
              <Content i18nKey='pages.Landing.section.dataImport.description' />
            </p>
          </div>
        </Section>
      )}

      {phase?.showBenefitsSections && (
        <>
          <Section id='more-benefits'>
            <h2>
              <Content i18nKey='pages.Landing.section.more-benefits.heading' />
            </h2>
            <div className='grid-row content-container'>
              <div className='grid-col'>
                <Content i18nKey='pages.Landing.section.more-benefits.benefits' />
              </div>
            </div>
            <div className='grid-row faq'>
              {/** First column of accordions */}
              {phase.showOpenSeasonFAQs && (
                <div className='grid-col'>
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.state-tax' />
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.learn-more' />
                </div>
              )}
              {phase.showEndOfFilingFAQs && (
                <div className='grid-col'>
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.need-to-file' />
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.amend-return' />
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.learn-more' />
                </div>
              )}
              {phase.showPostFilingSeasonFAQs && (
                <div className='grid-col'>
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.state-tax' />
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.learn-more' />
                </div>
              )}
              {/** Second column of accordions */}
              {phase.showOpenSeasonFAQs && (
                <div className='grid-col'>
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.tax-questions' />
                  {phase.showDataImportContent && (
                    <DFAccordion
                      splash
                      i18nKey='pages.Landing.section.faq.data-import'
                      allowedTags={[`p`, `h4`, `ul`, `li`]}
                    />
                  )}
                </div>
              )}
              {phase.showEndOfFilingFAQs && (
                <div className='grid-col'>
                  <DFAccordion
                    splash
                    i18nKey='pages.Landing.section.faq.state-tax-maybe-late'
                    allowedTags={[`p`, `h4`]}
                  />
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.rejected-still-time' />
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.where-is-refund' />
                  {phase.showDataImportContent && (
                    <DFAccordion
                      splash
                      i18nKey='pages.Landing.section.faq.data-import'
                      allowedTags={[`p`, `h4`, `ul`, `li`]}
                    />
                  )}
                </div>
              )}
              {phase.showPostFilingSeasonFAQs && (
                <div className='grid-col'>
                  <DFAccordion splash i18nKey='pages.Landing.section.faq.when-open' />
                  {phase.showDataImportContent && (
                    <DFAccordion
                      splash
                      i18nKey='pages.Landing.section.faq.data-import'
                      allowedTags={[`p`, `h4`, `ul`, `li`]}
                    />
                  )}
                </div>
              )}
            </div>
          </Section>
          <Section id='benefits'>
            <h2>
              <Content i18nKey='pages.Landing.section.benefits.heading' />
            </h2>
            <div className='feature-list'>
              <div className='container'>
                <div className='margin-right-2'>
                  <ChatIcon />
                </div>
                <div className='flex-5'>
                  <p>
                    <Content i18nKey='pages.Landing.section.benefits.quote-1' />
                  </p>
                </div>
              </div>
              <div className='container tablet:margin-left-2 desktop:grid-offset-1'>
                <div className='margin-right-2'>
                  <ChatIcon />
                </div>
                <div className='flex-5'>
                  <p>
                    <Content i18nKey='pages.Landing.section.benefits.quote-2' />
                  </p>
                </div>
              </div>
            </div>
          </Section>
        </>
      )}
      {phase?.showAfterDeadlineSections && (
        <>
          <Section id='after-deadline'>
            <h2>
              <Content i18nKey='pages.Landing.section.after-deadline.heading' />
            </h2>
            <p className='margin-bottom-5'></p>
            <div className='grid-row after-deadline grid-gap'>
              {/** First column of accordions */}
              <div className='grid-col'>
                <DFAccordion i18nKey='pages.Landing.section.after-deadline.get-copies-of-return' />
                <DFAccordion i18nKey='pages.Landing.section.after-deadline.amend-return' />
              </div>
              <div className='grid-col'>
                <DFAccordion i18nKey='pages.Landing.section.after-deadline.rejected-return' />
                <DFAccordion i18nKey='pages.Landing.section.after-deadline.disaster-postponed-deadline' />
              </div>
            </div>
          </Section>
          <Section id='help-with-taxes'>
            <h2>
              <Content i18nKey='pages.Landing.section.help-with-taxes.heading' />
            </h2>
            <p className='margin-bottom-5'></p>
            <div className='grid-row help-with-taxes grid-gap'>
              <div className='grid-col'>
                <DFAccordion i18nKey='pages.Landing.section.help-with-taxes.where-is-refund' />
                <DFAccordion i18nKey='pages.Landing.section.help-with-taxes.get-help-with-taxes' />
              </div>
              <div className='grid-col'>
                <DFAccordion i18nKey='pages.Landing.section.help-with-taxes.other-ways-to-file' />
                <DFAccordion i18nKey='pages.Landing.section.help-with-taxes.submission-after-deadline' />
              </div>
            </div>
          </Section>
        </>
      )}
    </>
  );
};

export default Intro;
