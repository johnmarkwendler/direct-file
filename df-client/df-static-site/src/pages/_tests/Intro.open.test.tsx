import { render } from '@testing-library/react';

import Intro from '../Intro.js';
import { wrapComponent } from '../../test/helpers.js';

const mockDate: Date = vi.hoisted(() => new Date(2024, 1, 22, 0, 0, 0));
vi.mock(`../../constants.js`, async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...(actual as object),
    __esModule: true,
    TODAY: mockDate,
  };
});

describe(`Intro page`, () => {
  test(`matches snapshot`, () => {
    expect(mockDate.toLocaleString(`default`, { month: `long` })).toBe(`February`);
    const { asFragment } = render(wrapComponent(<Intro />));
    const fragment = asFragment();
    expect(fragment).toMatchSnapshot();
  });
});
