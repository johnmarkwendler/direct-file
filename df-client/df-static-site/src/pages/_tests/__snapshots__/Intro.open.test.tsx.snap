// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Intro page > matches snapshot 1`] = `
<DocumentFragment>
  <section
    id="welcome"
  >
    <div
      class="grid-container"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        aria-relevant="additions"
        role="log"
      >
        <h1 />
      </div>
      <div
        class="grid-row"
      >
        <div
          class="grid-col-12 desktop:grid-col-6 tablet:grid-col-6"
        >
          <p
            class="lead-in"
          />
          <p
            class="lead-in margin-bottom-1"
          />
          <ul
            class="grid-row usa-icon-list"
          >
            <li
              class="grid-col-6 usa-icon-list__item"
            >
              <div
                aria-hidden="true"
                class="text-blue usa-icon-list__icon"
              >
                <svg
                  class="usa-icon"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                  />
                </svg>
              </div>
              <div
                class="usa-icon-list__content"
              />
            </li>
            <li
              class="grid-col-6 usa-icon-list__item"
            >
              <div
                aria-hidden="true"
                class="text-blue usa-icon-list__icon"
              >
                <svg
                  class="usa-icon"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                  />
                </svg>
              </div>
              <div
                class="usa-icon-list__content"
              />
            </li>
            <li
              class="grid-col-6 usa-icon-list__item"
            >
              <div
                aria-hidden="true"
                class="text-blue usa-icon-list__icon"
              >
                <svg
                  class="usa-icon"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                  />
                </svg>
              </div>
              <div
                class="usa-icon-list__content"
              />
            </li>
            <li
              class="grid-col-6 usa-icon-list__item"
            >
              <div
                aria-hidden="true"
                class="text-blue usa-icon-list__icon"
              >
                <svg
                  class="usa-icon"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                  />
                </svg>
              </div>
              <div
                class="usa-icon-list__content"
              />
            </li>
          </ul>
          <p
            class="lead-in"
          />
        </div>
        <div
          class="grid-col-12 desktop:grid-col-6 tablet:grid-col-6 margin-top-5 hidden-on-mobile"
        >
          <img
            alt=""
            class="padding-left-05"
            src="/imgs/blueperson_laptop.png"
          />
        </div>
      </div>
      <div
        class="grid-row grid-gap"
      >
        <div
          class="grid-col-12 desktop:grid-col-8 margin-bottom-2"
        >
          <a
            class="usa-button"
            href="/state"
          >
            pages.Landing.button.text
          </a>
        </div>
      </div>
    </div>
  </section>
  <section
    id="data-import"
  >
    <div
      class="grid-container"
    >
      <h2 />
      <div
        class="grid-row content-container"
      >
        <p />
      </div>
    </div>
  </section>
  <section
    id="more-benefits"
  >
    <div
      class="grid-container"
    >
      <h2 />
      <div
        class="grid-row content-container"
      >
        <div
          class="grid-col"
        />
      </div>
      <div
        class="grid-row faq"
      >
        <div
          class="grid-col"
        >
          <div>
            <div
              class="usa-accordion usa-accordion--bordered"
              data-testid="accordion"
            >
              <h3
                class="usa-accordion__heading"
              >
                <button
                  aria-controls="pages.Landing.section.faq.state-tax"
                  aria-expanded="false"
                  class="usa-accordion__button"
                  data-testid="accordionButton_pages.Landing.section.faq.state-tax"
                  type="button"
                />
              </h3>
              <div
                class="usa-accordion__content usa-prose"
                data-testid="accordionItem_pages.Landing.section.faq.state-tax"
                hidden=""
                id="pages.Landing.section.faq.state-tax"
              />
            </div>
          </div>
          <div>
            <div
              class="usa-accordion usa-accordion--bordered"
              data-testid="accordion"
            >
              <h3
                class="usa-accordion__heading"
              >
                <button
                  aria-controls="pages.Landing.section.faq.learn-more"
                  aria-expanded="false"
                  class="usa-accordion__button"
                  data-testid="accordionButton_pages.Landing.section.faq.learn-more"
                  type="button"
                />
              </h3>
              <div
                class="usa-accordion__content usa-prose"
                data-testid="accordionItem_pages.Landing.section.faq.learn-more"
                hidden=""
                id="pages.Landing.section.faq.learn-more"
              />
            </div>
          </div>
        </div>
        <div
          class="grid-col"
        >
          <div>
            <div
              class="usa-accordion usa-accordion--bordered"
              data-testid="accordion"
            >
              <h3
                class="usa-accordion__heading"
              >
                <button
                  aria-controls="pages.Landing.section.faq.tax-questions"
                  aria-expanded="false"
                  class="usa-accordion__button"
                  data-testid="accordionButton_pages.Landing.section.faq.tax-questions"
                  type="button"
                />
              </h3>
              <div
                class="usa-accordion__content usa-prose"
                data-testid="accordionItem_pages.Landing.section.faq.tax-questions"
                hidden=""
                id="pages.Landing.section.faq.tax-questions"
              />
            </div>
          </div>
          <div>
            <div
              class="usa-accordion usa-accordion--bordered"
              data-testid="accordion"
            >
              <h3
                class="usa-accordion__heading"
              >
                <button
                  aria-controls="pages.Landing.section.faq.data-import"
                  aria-expanded="false"
                  class="usa-accordion__button"
                  data-testid="accordionButton_pages.Landing.section.faq.data-import"
                  type="button"
                />
              </h3>
              <div
                class="usa-accordion__content usa-prose"
                data-testid="accordionItem_pages.Landing.section.faq.data-import"
                hidden=""
                id="pages.Landing.section.faq.data-import"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section
    id="benefits"
  >
    <div
      class="grid-container"
    >
      <h2 />
      <div
        class="feature-list"
      >
        <div
          class="container"
        >
          <div
            class="margin-right-2"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="44"
              viewBox="0 0 50 44"
              width="50"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.83134 2H43.1608C45.7212 2 47.8152 4.15 47.8152 6.79V27.14C47.8152 29.77 45.7212 31.93 43.1608 31.93H38.1133C37.9853 31.93 37.8756 32.06 37.8756 32.22L37.9121 42.01C37.9121 42.01 28.32 42.01 28.1462 32.2C28.1462 32.06 28.0456 31.94 27.9176 31.94H6.8222C4.21613 31.94 2.0947 29.79 2.0947 27.15V6.8C2.0947 4.17 4.22527 2.01 6.8222 2.01L6.83134 2Z"
                fill="white"
                stroke="#1B1B1B"
                stroke-width="2.4"
              />
              <path
                d="M10.334 13.4502H32.2798H10.334Z"
                fill="white"
              />
              <path
                d="M10.334 13.4502H32.2798"
                stroke="#1B1B1B"
                stroke-linecap="square"
                stroke-width="2.4"
              />
              <path
                d="M10.3336 21.4502H39.5947H10.3336Z"
                fill="white"
              />
              <path
                d="M10.3336 21.4502H39.5947"
                stroke="#1B1B1B"
                stroke-linecap="square"
                stroke-width="2.4"
              />
            </svg>
          </div>
          <div
            class="flex-5"
          >
            <p />
          </div>
        </div>
        <div
          class="container tablet:margin-left-2 desktop:grid-offset-1"
        >
          <div
            class="margin-right-2"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="44"
              viewBox="0 0 50 44"
              width="50"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.83134 2H43.1608C45.7212 2 47.8152 4.15 47.8152 6.79V27.14C47.8152 29.77 45.7212 31.93 43.1608 31.93H38.1133C37.9853 31.93 37.8756 32.06 37.8756 32.22L37.9121 42.01C37.9121 42.01 28.32 42.01 28.1462 32.2C28.1462 32.06 28.0456 31.94 27.9176 31.94H6.8222C4.21613 31.94 2.0947 29.79 2.0947 27.15V6.8C2.0947 4.17 4.22527 2.01 6.8222 2.01L6.83134 2Z"
                fill="white"
                stroke="#1B1B1B"
                stroke-width="2.4"
              />
              <path
                d="M10.334 13.4502H32.2798H10.334Z"
                fill="white"
              />
              <path
                d="M10.334 13.4502H32.2798"
                stroke="#1B1B1B"
                stroke-linecap="square"
                stroke-width="2.4"
              />
              <path
                d="M10.3336 21.4502H39.5947H10.3336Z"
                fill="white"
              />
              <path
                d="M10.3336 21.4502H39.5947"
                stroke="#1B1B1B"
                stroke-linecap="square"
                stroke-width="2.4"
              />
            </svg>
          </div>
          <div
            class="flex-5"
          >
            <p />
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
